RepositoryCompiler/CodeModel/CodeParsers/CSharp/CaDETMemberMetricCalculator.cs

```diff
@@ -121,5 +137,404 @@ private int CountOccurrences(string text, string pattern)
+        }
+
+        private int CountComparisonOperators(MemberDeclarationSyntax method)
+        {
+            return method.DescendantNodes().OfType<BinaryExpressionSyntax>()
+                .Count(n => n.IsKind(SyntaxKind.EqualsExpression) || 
+                            n.IsKind(SyntaxKind.NotEqualsExpression) ||
+                            n.IsKind(SyntaxKind.LessThanExpression) ||
+                            n.IsKind(SyntaxKind.LessThanOrEqualExpression) ||
+                            n.IsKind(SyntaxKind.GreaterThanExpression) ||
+                            n.IsKind(SyntaxKind.GreaterThanOrEqualExpression));
+        }
+
+        private int CountMethodInvocations(MemberDeclarationSyntax method)
+        {
+            return method.DescendantNodes().OfType<InvocationExpressionSyntax>().Count();
+        }
+
+        private int CountUniqueMethodInvocations(MemberDeclarationSyntax method)
+        {
+            List<string> uniqueInvokedMethods = new List<string>();
+            var allInvokedMethods = method.DescendantNodes().OfType<InvocationExpressionSyntax>();
+            foreach (var invokedMethod in allInvokedMethods)
+            {
+                if (!uniqueInvokedMethods.Contains(invokedMethod.Expression.ToString()))
+                {
+                    uniqueInvokedMethods.Add(invokedMethod.Expression.ToString());
+                }
+            }
+            return uniqueInvokedMethods.Count();
+        }
+
+        private int CountNumberOfAssignments(MemberDeclarationSyntax method)
+        {
+            return method.DescendantNodes().OfType<AssignmentExpressionSyntax>().Count();
+        }
+
+        private int CountNumberOfNumericLiterals(MemberDeclarationSyntax method)
+        {
+            return method.DescendantNodes().OfType<LiteralExpressionSyntax>().Count(n => n.IsKind(SyntaxKind.NumericLiteralExpression));
+        }
+
+        private int CountNumberOfStringLiterals(MemberDeclarationSyntax method)
+        {
+            return method.DescendantNodes().OfType<LiteralExpressionSyntax>().Count(n => n.IsKind(SyntaxKind.StringLiteralExpression));
+        }
+
+        private int CountNumberOfMathOperations(MemberDeclarationSyntax method)
+        {
+            int count = CountBinaryExpressions(method);
+            count += CountUnaryExpressions(method);
+            return count;
+        }
+
+        private int CountBinaryExpressions(MemberDeclarationSyntax method)
+        {
+            return method.DescendantNodes().OfType<BinaryExpressionSyntax>()
+                .Count(n => n.IsKind(SyntaxKind.AddExpression) ||
+                            n.IsKind(SyntaxKind.SubtractExpression) ||
+                            n.IsKind(SyntaxKind.MultiplyExpression) ||
+                            n.IsKind(SyntaxKind.DivideExpression) ||
+                            n.IsKind(SyntaxKind.ModuloExpression) ||
+                            n.IsKind(SyntaxKind.LeftShiftExpression) ||
```

Comment:
The shift operators are counted as math operations?

Reply:
That is how it was implemented in the https://github.com/mauricioaniche/ck, so I followed that example.